src/utils/scheduler.h

```diff
@@ -0,0 +1,176 @@
+#pragma once
+
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <queue>
+#include <thread>
+#include <unordered_set>
+
+namespace util {
+
+template <typename Clock> class PeriodicScheduler;
+
+namespace impl {
+template <typename Clock> void notifyScheduler(PeriodicScheduler<Clock>& scheduler);
+}
+
+/**
+ * @brief Uses a single thread to periodically invoke callbacks at a given frequency.
+ *
+ * The underlying thread is initialized upon construction,
+ * so do not create PeriodicSchedulers that will not be used.
+ *
+ * @tparam Clock The clock to use for scheduling.
+ */
+template <typename Clock = std::chrono::steady_clock> class PeriodicScheduler {
+public:
+	/**
+	 * @brief The type of event ids.
+	 */
+	using eventid_t = uint64_t;
+
+	/**
+	 * @brief Create a new PeriodicScheduler.
+	 */
+	PeriodicScheduler()
+		: newEventAdded(false), quitting(false), nextID(0),
+		  thread(std::bind(&PeriodicScheduler::threadFn, this)) {}
+
+	/**
+	 * @brief Join the thread and destruct.
+	 */
+	~PeriodicScheduler() {
+		{
+			std::unique_lock lock(scheduleMutex);
+			quitting = true;
+			newEventAdded = true;
```

Comment:
Why is newEventAdded set to true in the destructor?

Reply:
The condition variable is being notified, so we need to set the associated flag